### PR TITLE
[DependencyInjection] Use glob pattern to load config files

### PR DESCRIPTION
**Pull request by pierredup**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21173
| License       | MIT
| Doc PR        | -

This relates to #21173, but I'm not sure if it completely fixes the issue.

This allows to use a glob pattern to load config files, which makes the following possible:

```
# config.yml
imports:
    - { resource: "*.yml" }
    - { resource: "folder/*.yml" }
    - { resource: "/etc/myapp/*.{yml,xml}" }
```

It can also be used in a container extension, if a bundle uses a lot of configs:

```
$loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__ . '/../Resources/config'));
$loader->load('*.yml');
$loader->load('routing/*');
```

### CODE REVIEW

File: src/Symfony/Component/Config/FileLocator.php [hunk 1]
```diff
@@ -42,11 +42,15 @@ public function locate($name, $currentPath = null, $first = true)
         }
 
         if ($this->isAbsolutePath($name)) {
-            if (!file_exists($name)) {
+            if (array() === $files = glob($name, GLOB_BRACE)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/FileLocator.php]: missing check for GLOB_BRACE, because it's is not always defined (see php.net/glob )

File: src/Symfony/Component/Config/FileLocator.php [hunk 2]
```diff
@@ -58,19 +62,28 @@ public function locate($name, $currentPath = null, $first = true)
...
-                    return $file;
-                }
+        $load = function ($file) use (&$filepaths) {
+            if (@file_exists($file)) {
                 $filepaths[] = $file;
             }
+        };
+
+        foreach ($paths as $path) {
+            if (array() !== $files = glob($path.DIRECTORY_SEPARATOR.$name)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/FileLocator.php]: no BLOG_BRACE here?

File: src/Symfony/Component/Config/FileLocator.php [hunk 3]
```diff
@@ -58,19 +62,28 @@ public function locate($name, $currentPath = null, $first = true)
...
+                array_walk($files, $load);
+            } else {
+                $load($path.DIRECTORY_SEPARATOR.$name);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/FileLocator.php]: Does this actually do something? Looks like "glob" failed to find any file, so why would "file_exists" (in $load) return something else that "false"? If I'm right, then I'd prefer this array_walk + $load to be replaced by a simple foreach (with no `array() !== ...` check). But I may miss something, that's why I'm asking :)

Comment (nicolas-grekas) [inline src/Symfony/Component/Config/FileLocator.php]: (if it's legit, we should be sure it has a test case btw)

File: src/Symfony/Component/Config/FileLocator.php (hunk 1 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: Check added

File: src/Symfony/Component/Config/FileLocator.php (hunk 2 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: GLOB_BRACE added with a defined check

File: src/Symfony/Component/Config/FileLocator.php (hunk 3 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: I agree that this doesn't do anything.
Originally I wanted to preserve any existing functionality, but the `glob` should cover all current use cases.

File: src/Symfony/Component/Config/FileLocator.php [hunk 4]
```diff
@@ -42,11 +42,15 @@ public function locate($name, $currentPath = null, $first = true)
         }
 
         if ($this->isAbsolutePath($name)) {
-            if (!file_exists($name)) {
+            if (array() === $files = glob($name, defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
                 throw new FileLocatorFileNotFoundException(sprintf('The file "%s" does not exist.', $name));
             }
 
-            return $name;
+            if ($first) {
```
Comment (GuilhemN) [inline src/Symfony/Component/Config/FileLocator.php]: Wdyt about adding `|| 1 === count($files)` to keep the return the same as now?

File: src/Symfony/Component/Config/FileLocator.php [hunk 5]
```diff
@@ -59,10 +63,7 @@ public function locate($name, $currentPath = null, $first = true)
         $filepaths = array();
 
         foreach ($paths as $path) {
-            if (@file_exists($file = $path.DIRECTORY_SEPARATOR.$name)) {
-                if (true === $first) {
-                    return $file;
-                }
+            foreach (glob($path.DIRECTORY_SEPARATOR.$name, defined('GLOB_BRACE') ? GLOB_BRACE : 0) as $file) {
```
Comment (GuilhemN) [inline src/Symfony/Component/Config/FileLocator.php]: `array_merge`?

File: src/Symfony/Component/Config/FileLocator.php (hunk 4 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: Actually I think we should keep the original behavior as it was (just invert the `file_exists` check), then add the glob as an extra step.
Otherwise if you do something like `/etc/myapp/*.yml`, you would expect an array, so it might be weird to get a string if the glob only matches a single file

File: src/Symfony/Component/Config/FileLocator.php (hunk 5 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: Done

File: src/Symfony/Component/Config/FileLocator.php (hunk 4 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: Changes made

File: src/Symfony/Component/Config/FileLocator.php [hunk 6]
```diff
@@ -42,11 +42,19 @@ public function locate($name, $currentPath = null, $first = true)
...
+            if (file_exists($name)) {
+                return $name;
+            }
+
+            if (array() === $files = glob($name, defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
                 throw new FileLocatorFileNotFoundException(sprintf('The file "%s" does not exist.', $name));
             }
 
-            return $name;
+            if (true === $first) {
```
Comment (ro0NL) [inline src/Symfony/Component/Config/FileLocator.php]: Could be qualified a bugfix for 2.7 i guess, `$first` seems to be ignored previously :confused:

File: src/Symfony/Component/Config/FileLocator.php [hunk 7]
```diff
@@ -42,11 +42,19 @@ public function locate($name, $currentPath = null, $first = true)
...
+                return array_shift($files);
```
Comment (ro0NL) [inline src/Symfony/Component/Config/FileLocator.php]: I think `return reset($files)` is a faster operation..

File: src/Symfony/Component/Config/FileLocator.php [hunk 8]
```diff
@@ -59,18 +67,17 @@ public function locate($name, $currentPath = null, $first = true)
         $filepaths = array();
 
         foreach ($paths as $path) {
-            if (@file_exists($file = $path.DIRECTORY_SEPARATOR.$name)) {
-                if (true === $first) {
-                    return $file;
-                }
-                $filepaths[] = $file;
-            }
+            $filepaths = array_merge($filepaths, glob($path.DIRECTORY_SEPARATOR.$name, defined('GLOB_BRACE') ? GLOB_BRACE : 0));
```
Comment (ro0NL) [inline src/Symfony/Component/Config/FileLocator.php]: I'd prefer a quick return if `$first=true` and the first file is found.

File: src/Symfony/Component/Config/FileLocator.php (hunk 6 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: This would always have returned only one result, so the check for `$first` wasn't necessary. But the `glob` could return more than one result, so the the check needed to be added

File: src/Symfony/Component/Config/FileLocator.php (hunk 7 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: `array_shift` is used more commonly across the rest of the code base when doing a return. I don't think it makes much of a difference either way

File: src/Symfony/Component/Config/FileLocator.php (hunk 8 above)
Comment (pierredup) [inline src/Symfony/Component/Config/FileLocator.php]: Early return added

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 9]
```diff
@@ -138,7 +138,18 @@ private function parseImports($content, $file)
...
+            if (!$this->isAbsolutePath($import['resource'])) {
+                $import['resource'] = $defaultDirectory.DIRECTORY_SEPARATOR.$import['resource'];
+            }
+
+            if (array() !== $files = glob($import['resource'], defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
+                foreach ($files as $resource) {
+                    $this->import($resource, isset($import['type']) ? $import['type'] : null, isset($import['ignore_errors']) ? (bool) $import['ignore_errors'] : false, $file);
+                }
+            } else {
+                $this->import($import['resource'], isset($import['type']) ? $import['type'] : null, isset($import['ignore_errors']) ? (bool) $import['ignore_errors'] : false, $file);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: same question as FileLocator: is this "else" dead code or live code (same on the YamlFileLoader)

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 10]
```diff
@@ -106,7 +106,19 @@ private function parseImports(\DOMDocument $xml, $file)
...
-            $this->import($import->getAttribute('resource'), XmlUtils::phpize($import->getAttribute('type')) ?: null, (bool) XmlUtils::phpize($import->getAttribute('ignore-errors')), $file);
+
+            $resourceFile = $import->getAttribute('resource');
+            if (!$this->isAbsolutePath($resourceFile)) {
+                $resourceFile = $defaultDirectory.DIRECTORY_SEPARATOR.$resourceFile;
+            }
+
+            if (array() !== $files = glob($resourceFile, defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
+                foreach ($files as $resource) {
+                    $this->import($resource, XmlUtils::phpize($import->getAttribute('type')) ?: null, (bool) XmlUtils::phpize($import->getAttribute('ignore-errors')), $file);
```
Comment (sstok) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: The value of 'type' and 'ignore-errors' will not change within this loop, so it's better to resolve them before the loop, and then reuse the value.

**Review by sstok — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 9 above)
Comment (pierredup) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: I need to rethink this part a bit, because without this, you don't get the exception if a file doesn't exist

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php (hunk 10 above)
Comment (pierredup) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: Variables added for the two values

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 11]
```diff
@@ -106,7 +106,21 @@ private function parseImports(\DOMDocument $xml, $file)
         $defaultDirectory = dirname($file);
         foreach ($imports as $import) {
             $this->setCurrentDir($defaultDirectory);
-            $this->import($import->getAttribute('resource'), XmlUtils::phpize($import->getAttribute('type')) ?: null, (bool) XmlUtils::phpize($import->getAttribute('ignore-errors')), $file);
+
+            $resourceFile = $import->getAttribute('resource');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: Two notes here:
- tests still pass if I revert this change on XmlFileLoader on my setup
- I think this logic should be handled at the FileLoader level: I don't see why both Xml&Yaml loaders should care about that. They should be untouched to me.

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 12]
```diff
@@ -34,4 +34,15 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
 
         parent::__construct($locator);
     }
+
+    protected function isAbsolutePath($file)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: this looks like leaking logic from the locator, looks suspicious to me

**Comment by nicolas-grekas** [comment]
While working on using glob for another use case in #21289, I found some similarities with this PR.
Here is what I did there. It looks like this PR could do almost exactly the same:

- let the Config component untouched
- move the glob call to FileLoader in the DI component

Before calling glob, the "resource" should be parsed for glob-specific characters. The constant part of the "resource" should be the part that is given the locator. Then, FileLoader could use glob on the result, with the glob tail appended.

In case this is not clear enough, this logic is implemented here:
https://github.com/symfony/symfony/pull/21289/files#diff-ad1ed76aba6a80df5a48dfa4585adcf3R61

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 13]
```diff
@@ -34,4 +49,48 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
...
+        }
+
+        if (!$files = glob($directoryGlob, defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
+            throw new FileLoaderLoadException($directoryGlob, $sourceResource);
+        }
+
+        foreach ($files as $file) {
+            if (is_dir($file)) {
+                $this->container->addResource(new DirectoryResource($file, '/^$/'));
+                parent::import($file, 'directory', $ignoreErrors, $sourceResource);
```
Comment (pierredup) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: Directories should be handle by the DirectoryLoader class, otherwise we get an exception here.
Maybe we can deprecate the DirectoryLoader class in favor of using a glob?

**Comment by pierredup** [comment]
@nicolas-grekas I have made the suggested changes, please have a look if that is what you had in mind.

For me the downside of keeping this in the DI component and out of the Config component, is that you loose the ability to load multiple configs in a container extension, E.G

```
$loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__ . '/../Resources/config'));
$loader->load('*.yml');
```

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 14]
```diff
@@ -34,4 +49,42 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
...
+            $directoryGlob = $directoryPrefix.$directoryGlob;
+
+            if (!$files = glob($directoryGlob, defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
+                throw new FileLoaderLoadException($directoryGlob, $sourceResource);
+            }
+
+            foreach ($files as $file) {
+                if (is_dir($file)) {
+                    $this->container->addResource(new DirectoryResource($file, '/^$/'));
+                    parent::import($file, 'directory', $ignoreErrors, $sourceResource);
```
Comment (pierredup) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: Directories should be handle by the DirectoryLoader class, otherwise we get an exception here.
Maybe we can deprecate the DirectoryLoader class in favor of using a glob?

**Comment by pierredup** [comment]
The other issue with not changing the `FileLocator::locate`method, is that you now always need to have a directory prefix. You can't just load, for example, something like `*.yml` or`dev_*.yml`, and use whatever the paths is set in the locator

**Comment by pierredup** [comment]
The following patch fixes the above mentioned issues, wdyt about applying this changes?

```diff
 src/Symfony/Component/Config/FileLocator.php                   | 10 ++++++----
 .../Component/DependencyInjection/Loader/FileLoader.php        |  4 +++-
 2 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/src/Symfony/Component/Config/FileLocator.php b/src/Symfony/Component/Config/FileLocator.php
index 6b686390bc..09288b4bde 100644
--- a/src/Symfony/Component/Config/FileLocator.php
+++ b/src/Symfony/Component/Config/FileLocator.php
@@ -59,11 +59,13 @@ class FileLocator implements FileLocatorInterface
         $filepaths = array();

         foreach ($paths as $path) {
-            if (@file_exists($file = $path.DIRECTORY_SEPARATOR.$name)) {
-                if (true === $first) {
-                    return $file;
+            foreach(glob($path.DIRECTORY_SEPARATOR.$name, defined('GLOB_BRACE') ? GLOB_BRACE : 0) as $file) {
+                if (@file_exists($file)) {
+                    if (true === $first) {
+                        return $file;
+                    }
+                    $filepaths[] = $file;
                 }
-                $filepaths[] = $file;
             }
         }

diff --git a/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php b/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
index 46e89ab1e9..983cfbc5db 100644
--- a/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
+++ b/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
@@ -54,7 +54,9 @@ abstract class FileLoader extends BaseFileLoader
      */
     public function import($resource, $type = null, $ignoreErrors = false, $sourceResource = null)
     {
-        if (strlen($resource) === $i = strcspn($resource, '*?{[')) {
+        $i = strcspn($resource, '*?{[');
+
+        if (strlen($resource) === $i || 0 === $i || false === strpos($resource, '/')) {
             $directoryPrefix = $resource;
             $directoryGlob = '';
         } else {
```

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 15]
```diff
@@ -14,6 +14,9 @@
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\Config\Loader\FileLoader as BaseFileLoader;
 use Symfony\Component\Config\FileLocatorInterface;
+use Symfony\Component\Config\Exception\FileLoaderLoadException;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: should be moved before FileLoader (we add new entries in alpha order, but keep existing ones asis)

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 16]
```diff
@@ -34,4 +48,47 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
...
+            $directoryPrefix = realpath($directoryPrefix) ?: $directoryPrefix;
+            $directoryGlob = $directoryPrefix.$directoryGlob;
+
+            if (!$files = glob($directoryGlob, defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
+                throw new FileLoaderLoadException($directoryGlob, $sourceResource);
+            }
+
+            foreach ($files as $file) {
+                if (is_dir($file)) {
+                    $this->container->addResource(new DirectoryResource($file, '/^$/'));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: This misses tracking new/removed files in $directoryPrefix.
I think it's better to track at the prefix level as done at https://github.com/symfony/symfony/pull/21289/files#diff-ad1ed76aba6a80df5a48dfa4585adcf3R74:
`$this->container->addResource(new DirectoryResource($directoryPrefix, '/^$/'));`

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php (hunk 14 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: I think it's ok as is, no need to deprecate yet, that wouldn't provide much benefit, while forcing everyone to upgrade, isn't it?

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 17]
```diff
@@ -34,4 +48,47 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
...
+                    parent::import($file, 'directory', $ignoreErrors, $sourceResource);
+                } else {
+                    $this->container->addResource(new FileResource($file));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: resource tracking shouldn't be required here, import already does it (by calling load)

**Review by nicolas-grekas — COMMENTED** [review]
The current code looks good to me (with small comment).
I don't think FileLocator needs to be changed. Making it handle glob breaks it seeking logic to me. Better use the locator for the prefix, the glob inside the located result.

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php (hunk 17 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: Now that we have #21408, resource tracking should be done using the new `$container->fileExists()` method, both here and above instead of Directory/FileResource.

**Comment by nicolas-grekas** [comment]
Status: needs work

**Comment by pierredup** [comment]
@nicolas-grekas Changes made. I still think it's important to load config without the need of a directory prefix, but I'll work separately on that. If I can get something to work, I'll open a new PR

**Comment by pierredup** [comment]
Status: Needs Review

**Comment by nicolas-grekas** [comment]
Sure, when there is no prefix, we should just use ./

**Comment by pierredup** [comment]
That might not be intuitive for users, so it should probably then just be documented

**Comment by nicolas-grekas** [comment]
here is a fix - a separate test case would be better indeed
```diff
diff --git a/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php b/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
index 555831b..0f642dc 100644
--- a/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
+++ b/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
@@ -55,6 +55,9 @@ abstract class FileLoader extends BaseFileLoader
         if (strlen($resource) === $i = strcspn($resource, '*?{[')) {
             $directoryPrefix = $resource;
             $directoryGlob = '';
+        } elseif (0 === $i) {
+            $directoryPrefix = '.';
+            $directoryGlob = $resource;
         } else {
             $directoryPrefix = dirname(substr($resource, 0, 1 + $i));
             $directoryGlob = substr($resource, strlen($directoryPrefix));
diff --git a/src/Symfony/Component/DependencyInjection/Tests/Loader/FileLoaderTest.php b/src/Symfony/Component/DependencyInjection/Tests/Loader/FileLoaderTest.php
index ad875ba..8853eb6 100644
--- a/src/Symfony/Component/DependencyInjection/Tests/Loader/FileLoaderTest.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Loader/FileLoaderTest.php
@@ -43,7 +43,7 @@ class FileLoaderTest extends \PHPUnit_Framework_TestCase
         ));
 
         $loader->setResolver($resolver);
-        $loader->import('Fixtures/{xml,yaml}/services2.{yml,xml}');
+        $loader->import('{F}ixtures/{xml,yaml}/services2.{yml,xml}');
 
         $actual = $container->getParameterBag()->all();
         $expected = array(
```

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 18]
```diff
@@ -24,6 +25,17 @@
 {
     protected $container;
 
+    private $currentDir;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: can be made protected, and setCurrentDir be removed IMO (also required in my other PR)

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 19]
```diff
@@ -34,4 +46,43 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
...
+                parent::import($file, is_dir($file) ? 'directory' : $type, $ignoreErrors, $sourceResource);
+            }
+        } catch (\Exception $e) {
+            if (!$ignoreErrors) {
+                // prevent embedded imports from nesting multiple exceptions
+                if ($e instanceof FileLoaderLoadException) {
+                    throw $e;
+                }
+
+                throw new FileLoaderLoadException($resource, $sourceResource);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: `throw new FileLoaderLoadException($resource, $sourceResource, null, $e);`

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 20]
```diff
@@ -34,4 +46,43 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
...
+            if (!$files = glob($directoryGlob, defined('GLOB_BRACE') ? GLOB_BRACE : 0)) {
+                throw new FileLoaderLoadException($directoryGlob, $sourceResource);
+            }
+
+            $this->container->fileExists($directoryPrefix, '/^$/');
+
+            foreach ($files as $file) {
+                parent::import($file, is_dir($file) ? 'directory' : $type, $ignoreErrors, $sourceResource);
+            }
+        } catch (\Exception $e) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: is this catch block required?

**Comment by fabpot** [comment]
Coding standard should be fixed (cf. fabbot)

**Comment by nicolas-grekas** [comment]
I suggest waiting for #21289 before working on this PR - it will need a rebase and share code (new `glob` method)

**Comment by nicolas-grekas** [comment]
rebase unlocked, you can now use the new "glob" function:
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php#L95

If you happen to chosse to use "true" as second argument, then I invite you to remove this argument altogether because then it's the only value used in the code base (I tend to think we should do that.)

**Comment by nicolas-grekas** [comment]
Thinking again about this one, I'd now say it should not handle directories at all - only files (meaning the $recursive is required :) )

**Comment by pierredup** [comment]
Rebased and updated to use the new glob method.

> I'd now say it should not handle directories at all - only files

Does that mean I also don't need to change the `$type` to `directory` when a path is a directory? Or is that part still fine?

**Comment by nicolas-grekas** [comment]
I think that part should be removed yes: just skip directories.

**Comment by pierredup** [comment]
It seems like directory loading can't be skipped, it makes the current tests fail. So I think either `$recursive` then needs to be `true`, or we need to pass `directory` as the type if the path is a directory

**Comment by nicolas-grekas** [comment]
This should do it, WDYT?
```diff
diff --git a/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php b/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
index 794e242..41c9da6 100644
--- a/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
+++ b/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php
@@ -90,7 +90,7 @@ abstract class FileLoader extends BaseFileLoader
         $extRegexp = defined('HHVM_VERSION') ? '/\\.(?:php|hh)$/' : '/\\.php$/';
 
         foreach ($this->glob($resource, true, $prefixLen) as $path => $info) {
-            if (!preg_match($extRegexp, $path, $m) || !$info->isFile() || !$info->isReadable()) {
+            if (!preg_match($extRegexp, $path, $m) || !$info->isReadable()) {
                 continue;
             }
             $class = $namespace.ltrim(str_replace('/', '\\', substr($path, $prefixLen, -strlen($m[0]))), '\\');
@@ -112,6 +112,11 @@ abstract class FileLoader extends BaseFileLoader
     private function glob($resource, $recursive, &$prefixLen = null)
     {
         if (strlen($resource) === $i = strcspn($resource, '*?{[')) {
+            if (!$recursive) {
+                yield $resource => new \SplFileInfo($resource);
+
+                return;
+            }
             $resourcePrefix = $resource;
             $resource = '';
         } elseif (0 === $i) {
@@ -134,9 +139,11 @@ abstract class FileLoader extends BaseFileLoader
                 if ($recursive && is_dir($path)) {
                     $flags = \FilesystemIterator::SKIP_DOTS | \FilesystemIterator::FOLLOW_SYMLINKS;
                     foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($path, $flags)) as $path => $info) {
-                        yield $path => $info;
+                        if ($info->isFile()) {
+                            yield $path => $info;
+                        }
                     }
-                } else {
+                } elseif (is_file($path)) {
                     yield $path => new \SplFileInfo($path);
                 }
             }
@@ -155,7 +162,7 @@ abstract class FileLoader extends BaseFileLoader
         }
 
         foreach ($finder->followLinks()->in($resourcePrefix) as $path => $info) {
-            if (preg_match($regex, substr($path, $prefixLen))) {
+            if (preg_match($regex, substr($path, $prefixLen)) && $info->isFile()) {
                 yield $path => $info;
             }
         }
```

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 21]
```diff
@@ -41,6 +42,22 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
...
     /**
+     * {@inheritdoc}
+     */
+    public function import($resource, $type = null, $ignoreErrors = false, $sourceResource = null)
+    {
+        try {
+            foreach ($this->glob($resource, false) as $path => $info) {
+                parent::import($path, $type, $ignoreErrors, $sourceResource);
+            }
+        } catch (FileLocatorFileNotFoundException $exception) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: Should be named $e for consistency with the code base

**Comment by pierredup** [comment]
Changes made and tests passing again

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by fabpot** [comment]
Thank you @pierredup.
